homeassistant/components/schedule/__init__.py

```diff
@@ -0,0 +1,305 @@
+"""Support for schedules in Home Assistant."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from datetime import datetime, timedelta
+import itertools
+import logging
+from typing import Literal
+
+import voluptuous as vol
+
+from homeassistant.const import (
+    ATTR_EDITABLE,
+    CONF_ICON,
+    CONF_ID,
+    CONF_NAME,
+    SERVICE_RELOAD,
+    STATE_OFF,
+    STATE_ON,
+)
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+from homeassistant.helpers.collection import (
+    IDManager,
+    StorageCollection,
+    StorageCollectionWebsocket,
+    YamlCollection,
+    sync_entity_lifecycle,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.helpers.integration_platform import (
+    async_process_integration_platform_for_component,
+)
+from homeassistant.helpers.service import async_register_admin_service
+from homeassistant.helpers.storage import Store
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    ATTR_NEXT_EVENT,
+    CONF_ALL_DAYS,
+    CONF_FROM,
+    CONF_TO,
+    DOMAIN,
+    LOGGER,
+    WEEKDAY_TO_CONF,
+)
+
+STORAGE_VERSION = 1
+STORAGE_VERSION_MINOR = 1
+
+
+def has_no_overlap(schedule: list[dict[str, str]]) -> list[dict[str, str]]:
+    """Test the time ranges don't overlap."""
+    times = itertools.chain(
+        *[[time_range[CONF_FROM], time_range[CONF_TO]] for time_range in schedule]
+    )
+    if any(
+        time_range[CONF_FROM] < time < time_range[CONF_TO]
+        for time_range in schedule
+        for time in times
+    ):
+        raise vol.Invalid("Overlapping times found in schedule")
```

Comment:
let's add some comments to help future readers figure out what the code does
```suggestion
    # Make a list of all start and end times
    times = itertools.chain(
        *[[time_range[CONF_FROM], time_range[CONF_TO]] for time_range in schedule]
    )
    # Check if any time falls within the range of another time
    if any(
        time_range[CONF_FROM] < time < time_range[CONF_TO]
        for time_range in schedule
        for time in times
    ):
        raise vol.Invalid("Overlapping times found in schedule")
```

Reply:
The problem with this approach is that 2 exact overlapping periods will not be detected. You can verify that by updating the test to be:

```python
CONF_SUNDAY: [
    # {CONF_FROM: "00:00:00", CONF_TO: "23:59:59"},
    {CONF_FROM: "07:00:00", CONF_TO: "08:00:00"},
    {CONF_FROM: "07:00:00", CONF_TO: "08:00:00"},
],
```

A different approach to validate this would be to sort the schedule by `CONF_FROM` and making sure that the `CONF_FROM` always starts at or after the previous `CONF_TO`.

Reply:
I changed it to that approach, added the given example as a test case.
Added some more checks in the process.